progress/progress.go

```diff
@@ -53,147 +51,55 @@ type Meter interface {
-func (t *TextProgress) Set(current float64) {
-	t.pbar.Set(int(current))
-}
+func (NullMeter) Start(string, float64)       {}
+func (NullMeter) Set(float64)                 {}
+func (NullMeter) SetTotal(float64)            {}
+func (NullMeter) Finished()                   {}
+func (NullMeter) Write(p []byte) (int, error) { return len(p), nil }
+func (NullMeter) Notify(string)               {}
+func (NullMeter) Spin(msg string)             {}
 
-// SetTotal set the total steps needed
-func (t *TextProgress) SetTotal(total float64) {
-	t.pbar.Total = int64(total)
-}
-
-// Finished stops displaying the progress
-func (t *TextProgress) Finished() {
-	if t.pbar != nil {
-		// workaround silly pb that always does a fmt.Println() on
-		// finish (unless NotPrint is set)
-		t.pbar.NotPrint = true
-		t.pbar.Finish()
-		t.pbar.NotPrint = false
-	}
-	fmt.Printf("\r\033[K")
-}
+// QuietMeter is a Meter that _just_ shows Notify()s.
+type QuietMeter struct{ NullMeter }
 
-// Write is there so that progress can implment a Writer and can be
-// used to display progress of io operations
-func (t *TextProgress) Write(p []byte) (n int, err error) {
-	return t.pbar.Write(p)
+func (QuietMeter) Notify(msg string) {
+	fmt.Println(msg)
 }
 
-// Spin advances a spinner, i.e. can be used to show progress for operations
-// that have a unknown duration
-func (t *TextProgress) Spin(msg string) {
-	states := `|/-\`
+// testMeter, if set, is returned by MakeProgressBar; set it from tests.
+var testMeter Meter
 
-	// clear until end of line
-	fmt.Printf("\r[%c] %s%s", states[t.spinStep], msg, clearUntilEOL)
-	t.spinStep++
-	if t.spinStep >= len(states) {
-		t.spinStep = 0
+func MockMeter(meter Meter) func() {
+	testMeter = meter
+	return func() {
+		testMeter = nil
 	}
 }
 
-// Agreed asks the user whether they agree to the given license text
-func (t *TextProgress) Agreed(intro, license string) bool {
-	if _, err := fmt.Println(intro); err != nil {
-		return false
-	}
-
-	// XXX: send it through a pager instead of this ugly thing
-	if _, err := fmt.Println(license); err != nil {
-		return false
-	}
-
-	reader := bufio.NewReader(os.Stdin)
-	if _, err := fmt.Print("Do you agree? [y/n] "); err != nil {
-		return false
+var inTesting bool = len(os.Args) > 0 && strings.HasSuffix(os.Args[0], ".test") || os.Getenv("SPREAD_SYSTEM") != ""
+
+// MakeProgressBar creates an appropriate progress.Meter for the environ in
+// which it is called:
+//
+// * if MockMeter has been called, return that.
+// * if no terminal is attached, or we think we're running a test, a
+//   minimalistic QuietMeter is returned.
+// * otherwise, an ANSIMeter is returned.
+//
+// TODO: instead of making the pivot at creation time, do it at every call ( ....)
+var MakeProgressBar = func() Meter {
+	if testMeter != nil {
+		return testMeter
 	}
-	r, _, err := reader.ReadRune()
-	if err != nil {
-		return false
+	if !inTesting && terminal.IsTerminal(int(os.Stdin.Fd())) {
```

Comment:
Unrelated to this branch - but I wonder if we should have something `osutil.OnTerminal()` that automatically adds the zero fd. Anyway, not important here. We have at least one `Interactive: terminal.IsTerminal(0)` in the code already.

Reply:
we had `IsaTTY`, but dropped it in favour of this approach (because it used cgo, whereas this uses the syscall, but the syscall IIRC has arch shenanigans)